app/view_objects/allocations_view.rb

```diff
@@ -0,0 +1,57 @@
+class AllocationsView
+  module Status
+    REQUESTED = "REQUESTED".freeze
+    NOT_REQUESTED = "NOT REQUESTED".freeze
+    YET_TO_REQUEST = "YET TO REQUEST".freeze
+  end
+
+  module Colour
+    GREEN = "green".freeze
+    RED = "red".freeze
+    GREY = "grey".freeze
+  end
+
+  def initialize(training_providers:, allocations:)
+    @training_providers = training_providers
+    @allocations = allocations
+  end
+
+  def allocation_statuses
+    @training_providers.map do |training_provider|
+      matching_allocation = find_matching_allocation(training_provider)
+      build_allocation_status(matching_allocation, training_provider)
+    end
+  end
+
+private
+
+  def find_matching_allocation(training_provider)
+    @allocations.find do |allocation|
+      allocation.provider_id == training_provider.id.to_i
+    end
+  end
+
+  def build_allocation_status(matching_allocation, training_provider)
+    allocation_status = {
+      training_provider_name: training_provider.provider_name,
+      training_provider_code: training_provider.provider_code,
+    }
+
+    if matching_allocation.nil?
+      allocation_status[:status] = Status::YET_TO_REQUEST
+      allocation_status[:status_colour] = Colour::GREY
+    end
+
+    if matching_allocation && matching_allocation[:number_of_places] >= 1
+      allocation_status[:status] = Status::REQUESTED
+      allocation_status[:status_colour] = Colour::GREEN
+    end
+
+    if matching_allocation && matching_allocation[:number_of_places].zero?
+      allocation_status[:status] = Status::NOT_REQUESTED
+      allocation_status[:status_colour] = Colour::RED
+    end
+
+    allocation_status
+  end
```

Comment:
As part of this could we build the actions for each status could have? I'm thinking it would be good to refactor https://github.com/DFE-Digital/publish-teacher-training/blob/master/app/views/providers/allocations/_request_status.html.erb#L28-L53 and move logic out of the view and into this view_object.

Reply:
Hey @aliuk2012 - I've had have a think / play around with your suggestion and I'm not sure there is much scope for further refactoring (without introducing additional complexity). The key issue is that https://github.com/DFE-Digital/publish-teacher-training/blob/master/app/views/providers/allocations/_request_status.html.erb#L28-L53 now varies only in markup. The complex business logic has largely been removed and put into the Presenter (AllocationsView). It could be possible to make the AllocationsView generate markup but not sure if I am comfortable with this. 
A further option may be to extract the actions to an _actions_ partial if that would help?